src/constraints.jl

```diff
@@ -727,3 +727,313 @@ function list_of_constraint_types(model::Model)
+        end
+    end
+
+    if objective_sense(model) == MOI.MAX_SENSE
+        y .*= -1.0
+    end
+    return y
+end
+
+"""
+    perturbation_range_of_optimality(var::VariableRef)
+
+Gives the range by which the cost coefficient can change and the current lp-basis
+remains optimal, i.e., the reduced costs remain valid.
+
+## Notes
+- The range denote valid changes, Δ in [l, u], for which c[var] += Δ do not violate the current optimality conditions.
+
+"""
+function perturbation_range_of_optimality(var)
+    model = var.model
+    if termination_status(model) != MOI.OPTIMAL
+        error("The optimality range is not available because the current solution "*
+              "is not optimal.")
+    end
+    if objective_sense(model) == MOI.FEASIBILITY_SENSE
+        error("The optimality range is not applicable on feasibility problems.")
+    end
+    if !has_duals(model)
+        error("The optimality range is not available because no dual result is " *
+              "available.")
+    end
+    # Optimization possible: since the has_upper_bound(var) and has_lower_bound(var)
+    # Do not capture all variable bounds (issue 1892), it's hard to directly
+    # access the reduced cost of a variable, now the complete matrix is always built.
+
+    # The variable is in the basis for a minimization problem with no variable upper bounds we need:
+    # c_N - N^T B^-T (c_B + Δ e_j ) >= 0
+    # c_red - Δ N^T (B^-T e_j) = c_red - Δ N^T ρ = c_red - Δ N_red >= 0
+    # c_red >= Δ N_red
+    # c_red >= Δ N_red
+    # maximum( c_red_- ./ N_red_- ) <= Δ <= minimum( c_red_+ ./ N_red_+ )
+    # If upper bounds are present (and active), those inequalities are flipped.
+    # If the problem is a maximization problem all inequalities are flipped.
+    vars = all_variables(model)
+    j = findfirst(vars .== var)
+
+    A, L, U, constraints = _std_matrix(model)
+    var_basis_status = _std_basis_status(model)
+    c_red = _std_reduced_costs(model, constraints)
+
+    if var_basis_status[j] != MOI.BASIC
+        if L[j] == U[j]
+            return (-Inf, Inf)
+        end
+        if (var_basis_status[j] == MOI.NONBASIC_AT_LOWER && objective_sense(model) == MOI.MIN_SENSE) ||
+            (var_basis_status[j] == MOI.NONBASIC_AT_UPPER && objective_sense(model) == MOI.MAX_SENSE)
+            return (-c_red[j], Inf)
+        end
+        return (-Inf, -c_red[j])
+    end
+
+    basic = [var_basis_status[i] == MOI.BASIC for i in 1:length(var_basis_status)] # .== throws ERROR: MethodError: no method matching length(::MathOptInterface.BasisStatusCode)
+    upper = [var_basis_status[.!basic][i] == MOI.NONBASIC_AT_UPPER for i in 1:length(var_basis_status[.!basic])] # .== throws ERROR: MethodError: no method matching length(::MathOptInterface.BasisStatusCode)
+    ej = zeros(size(A)[1])
+    ej[findfirst(vars[basic[1:length(vars)]] .== var)] = 1.0
+    N_red = A[:, .!basic]' * (A[:, basic]' \ ej)
+    c_red = c_red[.!basic]
+    pos = N_red .> 1e-7
+    neg = N_red .< -1e-7
```

Comment:
Is `1e-7` special? What is the implication of a value that is `9e-8`?

Reply:
9e-8 is treated as zero. (neither negative nor positive)
Don't know if 1e-7 is the best choice. (But it usually works =) )
(This is related to the minimum ratio test in the simplex method.)
Choosing the epsilon too small makes the analysis break down, since a number that should be positive is due to numerical reasons deemed negative.  And too large epsilon is bad since elements that is clearly non-zero is regarded to be zero. 

But yes, making use of absolute tolerance assumes that the input is not strangely scaled...
Any ideas, experience, or other input are most welcomed.
(The most common argument for allowing the use of absolute tolerance within the simplex method is by first rescaling the columns and rows in A)